### PR TITLE
Added code to correctly handle HEAD request

### PR DESCRIPTION
**Pull request by dhoard**
Signed-off-by: Doug Hoard <doug.hoard@gmail.com>

### CODE REVIEW

**Comment by dhoard** [comment]
cc @fstab @tomwilkie

File: simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java [hunk 1]
```diff
@@ -84,6 +84,14 @@ protected ByteArrayOutputStream initialValue()
         public void handle(HttpExchange t) throws IOException {
             String query = t.getRequestURI().getRawQuery();
 
+            // Correct HEAD behavior per Java bug #6886723
+            if (t.getRequestMethod().equals("HEAD")) {
+                t.getRequestBody().close();
+                t.getResponseHeaders().add("Transfer-encoding", "chunked");
+                t.sendResponseHeaders(200, -1);
```
Comment (fstab) [inline simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java:91]: According to [the HTTP/1.1 Spec](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13)
> The Content-Length entity-header field indicates the size of the entity-body, in decimal number of OCTETs, sent to the recipient or, **in the case of the HEAD method, the size of the entity-body that would have been sent had the request been a GET.**

File: simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java [hunk 2]
```diff
@@ -84,6 +84,14 @@ protected ByteArrayOutputStream initialValue()
         public void handle(HttpExchange t) throws IOException {
             String query = t.getRequestURI().getRawQuery();
 
+            // Correct HEAD behavior per Java bug #6886723
+            if (t.getRequestMethod().equals("HEAD")) {
+                t.getRequestBody().close();
+                t.getResponseHeaders().add("Transfer-encoding", "chunked");
```
Comment (fstab) [inline simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java:90]: The `Content-type` header should be there as well. According to [the HTTP/1.1 Spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.4)
> The HEAD method is identical to GET except that the server MUST NOT return a message-body in the response. **The metainformation contained in the HTTP headers in response to a HEAD request SHOULD be identical to the information sent in response to a GET request.**

**Review by fstab — COMMENTED** [review]
Thanks a lot for the PR. I tried to reproduce the bug, but I could not find a single Java version where the `HEAD` request doesn't work. Even with Azul's Java 6 it works. Moreover, the current implementation handles headers correctly, so if I run
```
curl -i --head -H 'Accept: application/openmetrics-text; version=1.0.0; charset=utf-8' http://localhost:9000/metrics
```
I get the correct `Content-type` and `Content-length`.

Are you aware of any Java version where `HEAD` requests don't work? I would only fix it if it is a real problem.

**Comment by fstab** [comment]
I just saw that #685 is linked here, I commented there as well.

**Comment by dhoard** [comment]
I updated the PR. The issue is around the generation of the warning message. I have updated https://github.com/prometheus/client_java/issues/685 with specifics.

**Review by fstab — APPROVED** [review]
Ok, I see now what was the issue. Yes, it makes sense to fix this. The API of `sun.net.httpserver.HttpExchangeImpl.sendResponseHeaders()` is really strange, thanks for looking into it.

**Comment by yakirgb** [comment]
@fstab any option to cut new release?

**Pull request merged**